레스토랑 조회 구현하기 - heejae0704

Heejae0704 opened the pull request:
App 컴포넌트 기본형태에 대한 테스트를 작성하였습니다. (RED)

Heejae0704 committed:
App 기본형태 테스트 작성 (RED)

Heejae0704 committed:
App 컴포넌트 기본형태 구현 (GREEN)

Heejae0704 committed:
전체 위치 목록 불러오도록 리펙터 (REFACTOR)

Heejae0704 committed:
App 컴포넌트에서 Addresses 컴포넌트 분리 (REFACTOR)

Heejae0704 committed:
선택한 Address에 체크마크 표시 테스트 작성 (RED)

Heejae0704 committed:
선택된 Address에 체크마크 표시 구현 (GREEN)

Heejae0704 committed:
Addresses에서 AddressButton 컴포넌트 분리 (REFACTOR)

Heejae0704 committed:
Redux 사용할 준비작업 및 컴포넌트 이름 변경 (REFACTOR)

- Redux 사용을 위한 기본 세팅 작업을 구현
- API에서 사용하는 이름인 Region과 실제 자세한 주소를 의미하는 Address를 명확히
구분할 수 있도록 컴포넌트 이름들을 변경

Heejae0704 committed:
RegionsContainer 테스트 작성 (RED)

RegionsContainer가 redux store에서 regions 목록 가져오는지 테스트

Heejae0704 committed:
RegionsContainer 구현 (GREEN)

Heejae0704 committed:
markSelectedRegion action 구현 (RED->GREEN)

- region 버튼을 눌렀을 때 해당 버튼이 selected:true 값을 갖도록 구현
- 구현 방식을 고민하다보니 어느 정도 틀만 잡고 테스트를 하려다가 다 만들고 나서
테스트를 한 번에 통과해버려서.. RED->GREEN이 한 커밋에 있게 되었습니다.

heyjyu commented:
커밋을 한눈에 뭘 했는지 알 수 있게 매우 깔끔하게 잘 작성해주셨네요! TDD를 잘 하고 계신 것 같아요! 실무에서는 이렇게까지 잘게 커밋을 하지 않지만, 연습삼아 코드숨에서는 이렇게 의도적으로 하시면 도움이 많이 되실 것 같아요!

heyjyu commented on __mocks__/react-redux.js:
```diff
@@ -0,0 +1,2 @@
+export const useDispatch = jest.fn();
+export const useSelector = jest.fn();
```
이렇게 작성해도 큰 문제는 없지만
`useSelector를 jest.fn((selector) => selector({}))` 와 같이 모킹하면 조금 더 useSelector가 어떻게 쓰이는 지 알기 쉬울 것 같아요!

Heejae0704 replied:
혹시 두 방법의 차이점이 어떤 걸까요? 이렇게 구성해도 clearAllMocks를 해주어야 할까요?

heyjyu replied:
넵! 모킹을 할 때는 항상 클리어를 해주셔야 합니다~ 
mockImplementation을 하게 되면 결국 테스트에서 사용할 때 둘의 차이는 없습니다!
하지만 useSelector를 `jest.fn((selector) => selector({}))` 와 같이 모킹하면 useSelector는 selector라는 함수를 받아서 실행시키는 함수구나 하고 다른사람이 코드를 볼 때 짐작을 할 수 있게 됩니다!

heyjyu commented on src/App.test.jsx:
```diff
@@ -0,0 +1,41 @@
+  const dispatch = jest.fn();
+  useDispatch.mockImplementation(() => dispatch);
+
+  test('renders regions', () => {
```
describe을 사용하고 있다면 describe-context-it구조를 사용해주시면 더 테스트 spec을 확인하기 편리할 것 같아요!

Heejae0704 replied:
넵!

heyjyu commented on src/App.test.jsx:
```diff
@@ -0,0 +1,41 @@
+  return render(<App />);
+}
+
+describe('App', () => {
```
모킹을 이용하는 테스트에서는

```
  beforeEach(() => {
    jest.clearAllMocks();
  });
 ```
를 통해 모킹을 초기화해주는 게 좋습니다!

Heejae0704 replied:
넵!

heyjyu commented on fixtures/staticData.js:
```diff
@@ -0,0 +1,75 @@
+  {
+    id: 7,
+    name: '인천',
+    selected: false,
```
여러가지 방법이 있을 수 있지만 fixtures도 api와 구조를 맞춰주는 게 좋을 것 같아요!
https://eatgo-customer-api.ahastudio.com/regions

Heejae0704 replied:
네, fixture가 api와 구조가 같게 수정하고 특정 region 선택 시에 state의 region 정보와의 비교를 통해 (V)가 랜더링되도록 코드를 수정했습니다.

heyjyu commented on src/reducer.js:
```diff
@@ -0,0 +1,43 @@
+      regions: action.payload.regions,
+    };
+  }
+  if (action.type === 'markSelectedRegion') {
```
굳이 mark를 붙인 이유가 있을까요~? selectRegion으로도 원하는 의미를 충분히 담을 수 있을 것 같아서요!

Heejae0704 replied:
네 API 구조에 맞게 fixture 수정하면서 같이 고쳤습니다. 좀 더 고쳐서 region을 선택하는 내용과 category 선택하는 내용도 하나의 action으로 담을 수 있게 리펙터링할 예정입니다.

Heejae0704 replied:
이쪽 리펙터링을 시도하다가 오히려 더 복잡해지는 것 같아서 버튼 컴포넌트만 공유하는 수준에서 정리했습니다. ㅠ.ㅠ

heyjyu commented on src/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,29 @@
+  useDispatch.mockImplementation(() => dispatch);
+
+  test('renders regions', () => {
+    expect(screen.getByText('서울(V)')).not.toBeNull();
```
있는지 없는지 확인할 때는 getBy보다는 queryBy를 사용하는 게 좋아요! getBy는 찾지 못했을 때 에러를 발생시키기 때문입니다!

Heejae0704 replied:
넵!

Heejae0704 committed:
markSelectedRegion을 selectRegion으로 변경 (REFACTOR)

API와 fixture 틀을 맞추기 위해, fixture의 regions에 들어있던 selected 값을 사용하지
않고 region이라는 state를 별도로 만들고, 각 버튼 랜더링 시에 버튼 당 region의 이름과
region.name을 비교하여 (V)를 렌더링하는 방식으로 리펙터링.

리펙터링 후 깨진 테스트들을 수습함.

Heejae0704 committed:
CategoriesContainer 테스트 작성 (RED)

Heejae0704 committed:
CategoriesContainer 및 하위 컴포넌트 구현 (GREEN)

- 매우 많은 것들이 반복되기 때문에 리펙터링이 필요할 것이라는 점을 구현 중간에 깨달음
- 일단 정상 동작까지 시켜놓고 리펙터링하는 것으로 결정하고 구현 완료

TODO: 코드 반복이 심하기 때문에 공통되는 부분을 묶어주는 Generic한 컴포넌트로 리펙터링

Heejae0704 committed:
중복코드 RestaurantSelectorContainer로 리펙터링 테스트 작성(RED)

Heejae0704 committed:
RegionButton과 CategoryButton 컴포넌트 병합 (REFACTOR)

- 더 상위 레벨에서 리펙터링을 하려다가 더 복잡해지는 것을 깨닫고 포기
- 버튼 컴포넌트는 Region/Category와 관계없이 보편적으로 쓰일 수 있을 것으로 판단하여
  리펙터링

Heejae0704 committed:
현재까지 작업내용 테스트 커버리지 100% 되도록 수정 (REFACTOR)

heyjyu commented on __mocks__/react-redux.js:
```diff
@@ -0,0 +1,2 @@
+export const useDispatch = jest.fn();
+export const useSelector = jest.fn();
```
useSelector를 jest.fn((selector) => selector({})) 와 같이 모킹하면 조금 더 useSelector가 어떻게 쓰이는 지 알기 쉬울 것 같아요!

heyjyu commented on fixtures/staticData.js:
```diff
@@ -0,0 +1,66 @@
+  },
+];
+
+export const restaurants = [
```
https://github.com/CodeSoom/react-week5-assignment-1/pull/115#discussion_r937101661
restaurants도 api 와 구조를 맞춰주시면 좋을 것 같아요!

heyjyu commented on src/Regions.jsx:
```diff
@@ -0,0 +1,22 @@
+import SelectorButton from './SelectorButton';
+
+export default function Regions({ selectedRegionName, regions, onClick }) {
```
Categories와 Regions가 비슷한 점이 많아보이는 것 같아요! 어떻게 리팩토링 할 수 있을까요~?

Heejae0704 replied:
한 번 시도했다가 꼬여서 다시 한 번 도전해보려고 합니다.
저번에 실패했던 이유는.. 상위 컴포넌트 레벨의 리펙터링을 먼저 시도하다보니, action을 부르는 로직이 복잡해지는 부분 때문에 나중에 감당이 안되더라구요.

그래서 이번에는 컴포넌트들이 이용하는 하위 요소들인 actions, reducer, services 등을 먼저 공통적으로 사용할 수 있도록 모듈화를 하고, 상위 컴포넌트를 하나로 묶어주는 작업을 해보려고 합니다.

Heejae0704 replied:
둘을 RestaurantSelectorButtons 컴포넌트로 묶었습니다. 어쩌면 RegionsContainer와 CategoriesContainer도 묶을 수 있을 거 같은데... 들이는 노력에 비해 action, reducer 등의 대대적인 수정이 필요할 것 같아서 일단 이 작업에 대한 고민은 뒤로 미루려고 합니다.

heyjyu commented on src/reducer.test.js:
```diff
@@ -0,0 +1,88 @@
+    });
+  });
+
+  context('with no actions', () => {
```
with no 보다 without 을 사용하면 더 자연스러운 시나리오 작성을 할 수 있을 것 같습니다!

Heejae0704 replied:
넵!

heyjyu commented on src/reducer.test.js:
```diff
@@ -0,0 +1,88 @@
+    });
+  });
+
+  context('with not predefined action', () => {
```
https://github.com/CodeSoom/react-week5-assignment-1/pull/115#discussion_r937101661

heyjyu commented on src/reducer.test.js:
```diff
@@ -0,0 +1,88 @@
+
+  context('with setRestaurants', () => {
+    it('returns updated restaurants', () => {
+      expect(reducer(initialState, setRestaurants(restaurants))).toEqual({
```
```
 const state = reducer(initialState, setRestaurants(restaurants));
 expect(state.restaurants).toEqual(restaurants);
```
와 같이 작성하면 조금 더 확인하고 싶은 부분을 명확히 할 수 있을 것 같아요!

heyjyu commented on src/reducer.js:
```diff
@@ -0,0 +1,53 @@
+  region: {},
+  category: {},
+  regions,
+  categories,
```
initialState에 regions와 categories에 값이 들어 있지 않은 게 더 자연스러울 것 같은 데 혹시 어떻게 생각하시나요~?

Heejae0704 replied:
네 수정하였습니다! (이것 덕분에 fetch로 받아오는 작업 테스트가 꼬여서 한참을 고생했습니다 ^^;)

heyjyu commented on src/App.test.jsx:
```diff
@@ -0,0 +1,53 @@
+  });
+
+  context('with selected region and category', () => {
+    test('renders regions', () => {
```
describe-context-it을 활용하여 test를 it으로 수정해주시면 더 좋을 것 같아요!

Heejae0704 committed:
fetch로 categories와 regions 받아오기 구현

- actions 중에서 redux thunk 적용하는 부분 테스트 작성 (RED)
- actions 구현 (GREEN)
- static 데이터가 아닌 api에서 받아오는 값이 되도록 수정 (REFACTOR)

Heejae0704 committed:
Regions와 Categories 병함 테스트 작성 (RED)

Heejae0704 committed:
RestaurantSelectorButtons 구현 (GREEN)

Heejae0704 committed:
코드 리뷰 코멘트 반영 (REFACTOR)

Heejae0704 commented:
질문이 있습니다. `__mock__` 폴더에서 mocking을 해주는 것과, 해당 코드에서 직접 mocking하는 경우가 차이가 없다고 저번 리뷰에서 말씀해주셨는데요.. `action.js` 파일을 테스트하다보니 `action.test.js`에서 `const fetchRegions = jest.fn(() => regions)`로 쓰는 것과, `jest.mock('./services/api')`를 하는 것 간에 동작이 차이가 있어서 질문을 드리고 싶습니다.

`fetchRegions`는 `loadRegions`에서 불러서 쓰는 함수여서, 혹시 이런 경우에는 (아마도 함수 정의 시의 클로저라서?) 아예 모듈 자체를 mock하는 방식을 써야 하는 것인지.. 아니면 제가 뭔가 잘못 이해하고 있는 부분이 있는지.. 도움을 부탁드립니다!

test coverage 100%를 맞추려고 하다보니 actions에 있는 redux thunk에서 사용하는 비동기 함수의 테스트가 까다롭던데요... 이걸 뭔가 지금과 다르게 좀 더 쉽게 할 수 있는 방법이 있을까요...

Heejae0704 committed:
레스토랑 불러와서 렌더링하는 기능 테스트 작성 (RED)

Heejae0704 committed:
레스토랑 목록 불러오기 구현 (GREEN)

- E2E 테스트가 API 응답과 맞지 않아 통과되지 않습니다.

Heejae0704 commented:
일단 기능구현까지는 완료했는데요, API가 변경되어서인지 서울/한식 선택하면 양천주가가 나오는지를 테스트하는데 API 응답에는 양천주가가 없습니다.

Heejae0704 committed:
API mock 함수 수정

Heejae0704 committed:
아샬님 문제풀이 보고 나서 리펙터링

- reducer에서 reducers 객체 만들어서 action 이름에 따라 가져오는 패턴으로 변경

Heejae0704 committed:
ESLINT 정리

heyjyu commented:
> 질문이 있습니다. `__mock__` 폴더에서 mocking을 해주는 것과, 해당 코드에서 직접 mocking하는 경우가 차이가 없다고 저번 리뷰에서 말씀해주셨는데요.. `action.js` 파일을 테스트하다보니 `action.test.js`에서 `const fetchRegions = jest.fn(() => regions)`로 쓰는 것과, `jest.mock('./services/api')`를 하는 것 간에 동작이 차이가 있어서 질문을 드리고 싶습니다.
> 
> `fetchRegions`는 `loadRegions`에서 불러서 쓰는 함수여서, 혹시 이런 경우에는 (아마도 함수 정의 시의 클로저라서?) 아예 모듈 자체를 mock하는 방식을 써야 하는 것인지.. 아니면 제가 뭔가 잘못 이해하고 있는 부분이 있는지.. 도움을 부탁드립니다!
> 
> test coverage 100%를 맞추려고 하다보니 actions에 있는 redux thunk에서 사용하는 비동기 함수의 테스트가 까다롭던데요... 이걸 뭔가 지금과 다르게 좀 더 쉽게 할 수 있는 방법이 있을까요...

<img width="324" alt="image" src="https://user-images.githubusercontent.com/96414086/183243167-58ccdb1a-9358-40f1-9d6c-c194e96d36cc.png">
이 구조로 모킹을 하셨는 지 확인을 부탁드립니다!

heyjyu commented:
> 일단 기능구현까지는 완료했는데요, API가 변경되어서인지 서울/한식 선택하면 양천주가가 나오는지를 테스트하는데 API 응답에는 양천주가가 없습니다.

네 맞습니다.. 죄송합니다. 윤석님께 수정을 부탁드리겠습니다!

heyjyu commented:
> test coverage 100%를 맞추려고 하다보니 actions에 있는 redux thunk에서 사용하는 비동기 함수의 테스트가 까다롭던데요... 이걸 뭔가 지금과 다르게 좀 더 쉽게 할 수 있는 방법이 있을까요...

다른 라이브러리가 나온다면 모르겠지만, 지금은 redux-mock-store를 사용하는 작성하신 방식이 최선인 것 같습니다. 테스트를 잘 작성해주셨네요! 혹은 test coverage에 안잡히도록 무시할 수 있는 옵션이 있긴 하지만, 이 부분의 테스트는 필요한 내용이니 redux-mock-store을 이용하는 게 좋을 것 같습니다!

heyjyu commented:
> 질문이 있습니다. `__mock__` 폴더에서 mocking을 해주는 것과, 해당 코드에서 직접 mocking하는 경우가 차이가 없다고 저번 리뷰에서 말씀해주셨는데요.. `action.js` 파일을 테스트하다보니 `action.test.js`에서 `const fetchRegions = jest.fn(() => regions)`로 쓰는 것과, `jest.mock('./services/api')`를 하는 것 간에 동작이 차이가 있어서 질문을 드리고 싶습니다.
> 
> `fetchRegions`는 `loadRegions`에서 불러서 쓰는 함수여서, 혹시 이런 경우에는 (아마도 함수 정의 시의 클로저라서?) 아예 모듈 자체를 mock하는 방식을 써야 하는 것인지.. 아니면 제가 뭔가 잘못 이해하고 있는 부분이 있는지.. 도움을 부탁드립니다!

정확합니다.
services/api 자체를 모킹을 하거나, 만약에 테스트에서 모킹을 하시려고 한다면 loadRegions를 모킹을 해야합니다. 하지만 이것보다는 fetchRegions만 모킹을 하는 게 낫기 때문에 이 경우에는 `__mocks__` 폴더를 사용하시는 게 좋을 것 같습니다!
